### PR TITLE
Enable beatloop activation directly after activating a hotcue

### PR DESCRIPTION
**Pull request by goddisignz**
Fxes https://bugs.launchpad.net/mixxx/+bug/1464003 .
Basically the loopcontroller has to set the loop start position to the queued seek position before activating the loop.
Furthermore, when a hotcue that is at the end of an active loop is activated, also disable the loop instead of jumping back to the loop entry.

### CODE REVIEW

File: src/engine/controls/loopingcontrol.cpp [hunk 1]
```diff
@@ -958,6 +960,10 @@ void LoopingControl::updateBeatLoopingControls() {
 }
 
 void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable) {
+
+    // if a seek was queued in the engine buffer wait until it is executed
+    while(getEngineBuffer()->isSeekQueued()){}
```
Comment (uklotzde) [inline src/engine/controls/loopingcontrol.cpp]: This doesn't look sane to me. A busy wait on a condition that is controlled by another thread?

Comment (goddisignz) [reply src/engine/controls/loopingcontrol.cpp]: I expected something like this but was not sure which other mechanism to use. Just tell me what would me the right one and I will adapt my solution.

Comment (daschuer) [reply src/engine/controls/loopingcontrol.cpp]: A straight forward solution is to predict the effect of the scheduled seek here.

File: src/engine/controls/loopingcontrol.cpp [hunk 2]
```diff
@@ -782,7 +784,7 @@ void LoopingControl::notifySeek(double dNewPlaypos, bool adjustingPhase) {
             setLoopingEnabled(false);
         }
         if (currentSample <= loopSamples.end &&
-                dNewPlaypos > loopSamples.end) {
+                dNewPlaypos >= loopSamples.end) {
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: Why is this change necessary?

Comment (goddisignz) [reply src/engine/controls/loopingcontrol.cpp]: Say two hotcues (A and B) are 4 bars apart from each other. A 4 bar loop is active starting at hotcue A. The end of the loop is exactly at hotcue B. When hotcue B ist activated the loop should be deactivated and the position set to B. Without this change, the position is immediately set to the beginning of the loop because the loop is still active.

Does that clarify it?

Comment (daschuer) [reply src/engine/controls/loopingcontrol.cpp]: Yes, Thank you. Please explain this as a comment.

File: src/engine/controls/loopingcontrol.cpp (hunk 1 above)
Comment (daschuer) [reply src/engine/controls/loopingcontrol.cpp]: I think you need to make the 
LoopSamples loopSamples = m_loopSamples.getValue(); conditional. 
If there is a cued seek, you need to use m_queuedSeekPosition.getValue(). 

By luck you can ignore the quantize flag. If not you need to adjust the loop along with the   m_queuedSeekPosition.

File: src/engine/controls/loopingcontrol.cpp (hunk 2 above)
Comment (goddisignz) [reply src/engine/controls/loopingcontrol.cpp]: By comment you mean in the PR or in the code?

Comment (daschuer) [reply src/engine/controls/loopingcontrol.cpp]: Yes, right.

Comment (ronso0) [reply src/engine/controls/loopingcontrol.cpp]: .. in the code ;)

Comment (daschuer) [reply src/engine/controls/loopingcontrol.cpp]: :man_facepalming:  sorry

File: src/engine/controls/loopingcontrol.cpp [hunk 3]
```diff
@@ -958,6 +959,14 @@ void LoopingControl::updateBeatLoopingControls() {
 }
 
 void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable) {
+
+    // if a seek was queued in the engine buffer move the current sample to its position
+    ControlValueAtomic<double> seekPosition;
+    if(getEngineBuffer()->isSeekQueued(seekPosition)){
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: Please use git-clang-format HEAD~1 to fix some code style issues. 
https://www.mixxx.org/wiki/doku.php/coding_guidelines#command_line

File: src/engine/enginebuffer.cpp [hunk 4]
```diff
@@ -1163,12 +1163,24 @@ void EngineBuffer::processSeek(bool paused) {
...
+
+    if (syncPosition != m_filepos_play) {
+        if(seekType & SEEK_PHASE){
+            // inform controllers about the seek position itself
+            // and do the phase shift afterwards
+            // this prevents loops from getting disabled because the synced position is in front of the loop
+            setNewPlaypos(position, false);
+            adjustingPhase = true;
+        }
+        setNewPlaypos(syncPosition, adjustingPhase);
```
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: The double setNewPlaypos() call reads a bit like a hack. Can we add for example a new parameter to identify the situation explicit?

File: src/engine/enginebuffer.cpp [hunk 5]
```diff
@@ -1284,6 +1296,14 @@ bool EngineBuffer::isTrackLoaded() {
     return false;
 }
 
+bool EngineBuffer::isSeekQueued(ControlValueAtomic<double> &seekPos) {
```
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: we use only const references in Mixxx. I think we can use here a "double* pSeekPosition" as second return value.

File: src/engine/enginebuffer.cpp [hunk 6]
```diff
@@ -1284,6 +1296,14 @@ bool EngineBuffer::isTrackLoaded() {
...
+    if(!m_iSeekQueued){
```
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: Is this check redundant?

File: src/engine/controls/loopingcontrol.cpp [hunk 7]
```diff
@@ -958,6 +959,14 @@ void LoopingControl::updateBeatLoopingControls() {
 }
 
 void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable) {
+
+    // if a seek was queued in the engine buffer move the current sample to its position
+    ControlValueAtomic<double> seekPosition;
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: we can use here a plain double, because it lives on the stack and is not used concurrently.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for your update. 
I have added some comments. 

Does it work as expected now? 

It would be great to have a unit test for this case. 
Can you had one? Do you know how?

File: src/engine/enginebuffer.cpp (hunk 4 above)
Comment (goddisignz) [reply src/engine/enginebuffer.cpp]: Yep, this is more like a workaround but as I am still new to the project I didn't want to change parameters of already existing methods. We could have a seekedPosition and syncedPosition in setNewPlaypos and notifySeek and let the controller decide what to do with this information. How about that?

**Comment by goddisignz** [comment]
I will make changes and recommit.
It works as expected except in one case. If the synced position is in front of an active loop the playposition does not end up in front of the end of the loop. I would expect it to be like that.

File: src/engine/enginebuffer.cpp [hunk 8]
```diff
@@ -1163,12 +1163,24 @@ void EngineBuffer::processSeek(bool paused) {
...
+            // and do the phase shift afterwards
+            // this prevents loops from getting disabled because the synced position is in front of the loop
+            setNewPlaypos(position, false);
+            adjustingPhase = true;
+        }
+        setNewPlaypos(syncPosition, adjustingPhase);
     }
+
+    // seek finished after setNewPlaypos
+    m_iSeekQueued.fetchAndStoreRelaxed(SEEK_NONE);
```
Comment (uklotzde) [inline src/engine/enginebuffer.cpp]: Are you able to justify that the reordering of atomic memory operations still works as expected?

Comment (goddisignz) [reply src/engine/enginebuffer.cpp]: Actually, I couldn't. But it this line won't be necessary anymore.

File: src/engine/enginebuffer.cpp (hunk 6 above)
Comment (goddisignz) [reply src/engine/enginebuffer.cpp]: Yup, was.

**Review by daschuer — CHANGES_REQUESTED** [review]
Some more comments. We are getting close. Thank you.

File: src/engine/controls/loopingcontrol.cpp [hunk 9]
```diff
@@ -768,7 +769,7 @@ void LoopingControl::slotLoopEndPos(double pos) {
 }
 
 // This is called from the engine thread
-void LoopingControl::notifySeek(double dNewPlaypos, bool adjustingPhase) {
+void LoopingControl::notifySeek(double dNewPlaypos, double dSyncedNewPlayPos, bool adjustingPhase) {
```
[the remark this answers is not in the document]
Comment (goddisignz) [reply src/engine/controls/loopingcontrol.cpp]: Assume an active loop starting exactly at hotcue A. If I activate the hotcue the synced position is very likely in front of the loop. The adjustingPhase is still false, as it is was a SEEK_STANDARD_PHASE. If we set the adjustingPhase to true in that case, the loop would not be disabled for any other seek position outside the loop.
Thus, the loopcontroller should know about the position requested by the user but the remaining controllers should still be notified with the synced position.

I assume that the adjustPhase is still necessary. When there was no sync, both positions are equal. But (I guess) the same holds true if only a sync without any seeking is requested.

File: src/engine/enginebuffer.cpp [hunk 10]
```diff
@@ -1125,7 +1125,7 @@ void EngineBuffer::processSeek(bool paused) {
     // call anyway again.
 
     SeekRequests seekType = static_cast<SeekRequest>(
-            m_iSeekQueued.fetchAndStoreRelease(SEEK_NONE));
+            m_iSeekQueued.fetchAndStoreRelaxed(SEEK_NONE));
```
[the remark this answers is not in the document]
Comment (goddisignz) [reply src/engine/enginebuffer.cpp]: That must be an autocomplete-mistake. I wanted to change it back to the original version.

[the remark this answers is not in the document]
Comment (daschuer) [reply src/engine/enginebuffer.cpp]: But it was originally wrong. We need the Aquire memory fence here, to stop the pipeline executing the reading of the seek position too early.

File: src/engine/controls/loopingcontrol.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (goddisignz) [reply src/engine/controls/loopingcontrol.cpp]: I just had a thought on solving this case differently. Let the loopingcontroller calculate a position inside a loop (if enabled) based on the requested and the synced position.

```suggestion
    double syncPosition = position;
    if (!paused && (seekType & SEEK_PHASE)) {
        syncPosition = m_pBpmControl->getNearestPositionInPhase(position, true, true);
        position = m_pLoopingControl->getSyncPositionInsideLoop(position, syncPosition);
    }
```
The setNewPos and notifySeek could be left unchanged. Furthermore, a seek + sync that sets the position outside the loop can be corrected to be inside.

[the remark this answers is not in the document]
Comment (daschuer) [reply src/engine/controls/loopingcontrol.cpp]: Yes, even better readability. Thank you.

But why do we need syncPosition outside the if block? 
I think it would be even more clean to have a requested position as well and not reuse position for this.

File: src/engine/enginebuffer.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (goddisignz) [reply src/engine/enginebuffer.cpp]: OK, I will change that. But which value should I subtract?
Are you sure `fetchAndSubAcquire` shouldn't be `fetchAndStoreAcquire`?

[the remark this answers is not in the document]
Comment (daschuer) [reply src/engine/enginebuffer.cpp]: Ups ... of cause.

File: src/engine/controls/loopingcontrol.cpp [hunk 11]
```diff
@@ -806,6 +807,38 @@ bool LoopingControl::isLoopingEnabled() {
...
+    // the synced position is in front of the loop
+    // adjust the synced position to same amount in front of the loop end
+    if (dSyncedPlayPos <= loopSamples.start) {
+        return loopSamples.end - (loopSamples.start - dSyncedPlayPos);
+    }
+
+    // the synced position is behind the loop
+    // adjust the synced position to same amount behind the loop start
+    if (dSyncedPlayPos >= loopSamples.end) {
+        return loopSamples.start + (dSyncedPlayPos - loopSamples.end);
```
Comment (daschuer) [inline src/engine/controls/loopingcontrol.cpp]: the same here, this can still be behind the loop in case of small loops.

File: src/engine/enginebuffer.cpp [hunk 12]
```diff
@@ -1164,8 +1164,12 @@ void EngineBuffer::processSeek(bool paused) {
...
     if (!paused && (seekType & SEEK_PHASE)) {
-        position = m_pBpmControl->getNearestPositionInPhase(position, true, true);
+        double syncPosition = position;
+        double requestedPosition = position;
+        syncPosition = m_pBpmControl->getNearestPositionInPhase(position, true, true);
+        position = m_pLoopingControl->getSyncPositionInsideLoop(requestedPosition, syncPosition);
     }
+
     if (position != m_filepos_play) {
         setNewPlaypos(position, adjustingPhase);
```
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: i think we can now get rid of the adjustingPhase flag, because the position is always inside the loop now.

**Review by daschuer — CHANGES_REQUESTED** [review]
some more comments.

File: src/engine/controls/loopingcontrol.cpp [hunk 13]
```diff
@@ -806,6 +807,38 @@ bool LoopingControl::isLoopingEnabled() {
...
+    if (dRequestedPlaypos < loopSamples.start || dRequestedPlaypos >= loopSamples.end) {
+        return dSyncedPlayPos;
+    }
+
+    // the requested position is inside the loop (e.g hotcue at start)
+
+    // the synced position is in front of the loop
+    // adjust the synced position to same amount in front of the loop end
+    if (dSyncedPlayPos <= loopSamples.start) {
+        return loopSamples.end - (loopSamples.start - dSyncedPlayPos);
```
[the remark this answers is not in the document]
Comment (goddisignz) [reply src/engine/controls/loopingcontrol.cpp]: You're right, but not problem to solve. 
BTW: This whole method only works on beatloops (N-bar or fraction of N). Should we also care about loops which are non-beatloops or just ignore that case?

[the remark this answers is not in the document]
Comment (daschuer) [reply src/engine/controls/loopingcontrol.cpp]: I think it is independent from where loop in and out are set.

We may have this condition:
B____I__O
B=beat
I=loop in
O= loop out

We need only make sure that the time between. B and I is rolled of in the loop in a way that the requested point in the loop is reached as if the seek had been done to B.

So yes, we can ignore the case. The user is responsible to use a beat loop if he wishes the loop is in beat.

**Comment by goddisignz** [comment]
I will use create a new PR to fix https://bugs.launchpad.net/mixxx/+bug/1778246 in the 2.1 release.
The remaining changes will be put to a new PR
